services/src/main/java/org/keycloak/theme/ResourceLoader.java

```diff
@@ -12,19 +12,40 @@ public static InputStream getResourceAsStream(String root, String resource) thro
         if (root == null || resource == null) {
             return null;
         }
-        Path rootPath = Path.of("/", root).normalize().toAbsolutePath();
-        Path resourcePath = rootPath.resolve(resource).normalize().toAbsolutePath();
-        if (resourcePath.startsWith(rootPath)) {
-            if (File.separatorChar == '/') {
-                resource = resourcePath.toString().substring(1);
-            } else {
-                resource = resourcePath.toString().substring(2).replace('\\', '/');
-            }
-            URL url = classLoader().getResource(resource);
-            return url != null ? url.openStream() : null;
-        } else {
+        
+        // Check for encoded traversal patterns in the original resource
+        if (containsEncodedTraversal(resource)) {
```

Comment:
Do we expect any URL encoding in the resource string? If not, shouldn't we replace \\ with /, split on / and replace the inbetween parts with URL encoding? That should always preserve the users intention, and prevent the encoded traversal.